### PR TITLE
Migrating to optax: DQN and Rainbow

### PR DESCRIPTION
**Pull request by johanobandoc**
Migrating to optax: DQN and Rainbow:

1. add import optax
2. update _build_networks_and_optimizer function
3. add self.online_params and self.optimizer_state to call to train, and receive optimizer_state and online_params
4. change train signature so it accepts online_params and optimizer_state, and update the static_argnums in the jit decorator
5. update the way you apply the gradients (two lines instead of one)
6. return optimizer_state and online_params from train

### CODE REVIEW

File: lifting_veil/agents/dqn_agent_new.py [hunk 1]
```diff
@@ -31,17 +31,20 @@
...
-
+import optax
 
 def mse_loss(targets, predictions):
     return jnp.mean(jnp.power((targets - (predictions)), 2))
 
 
-@functools.partial(jax.jit, static_argnums=(0, 9, 10, 11, 12, 13, 14))
-def train(network_def, target_params, optimizer, states, actions, next_states, rewards, terminals, loss_weights,
+#@functools.partial(jax.jit, static_argnums=(0, 9, 10, 11, 12, 13, 14))
```
Comment (psc-g) [inline lifting_veil/agents/dqn_agent_new.py]: remove these commented out lines?

File: lifting_veil/agents/dqn_agent_new.py [hunk 2]
```diff
@@ -270,10 +274,10 @@ def __init__(self,
...
     def _build_networks_and_optimizer(self):
         self._rng, rng = jax.random.split(self._rng)
-        online_network_params = self.network_def.init(rng, x=self.state, rng=self._rng)
-        optimizer_def = dqn_agent.create_optimizer(self._optimizer_name)
-        self.optimizer = optimizer_def.create(online_network_params)
-        self.target_network_params = copy.deepcopy(online_network_params)
+        self.online_params = self.network_def.init(rng, x=self.state, rng=self._rng)
+        self.optimizer = create_optimizer(self._optimizer_name)
+        self.optimizer_state = self.optimizer.init(self.online_params)
+        self.target_network_params = copy.deepcopy(self.online_params)
```
Comment (psc-g) [inline lifting_veil/agents/dqn_agent_new.py:276]: you don't actually need the `copy.deepcopy` anymore

File: lifting_veil/agents/dqn_agent_new.py [hunk 3]
```diff
@@ -311,11 +315,24 @@ def _train_step(self):
...
-                self.optimizer, loss, mean_loss = train(
-                    self.network_def, self.target_network_params, self.optimizer, self.replay_elements['state'],
-                    self.replay_elements['action'], self.replay_elements['next_state'], self.replay_elements['reward'],
-                    self.replay_elements['terminal'], loss_weights, self.cumulative_gamma, self._target_opt,
-                    self._mse_inf, self._tau, self._alpha, self._clip_value_min, self._rng)
+                self.optimizer_state, self.online_params, loss = train(
+                    self.network_def,
+                    self.online_params,
+                    self.target_network_params,
+                    self.optimizer_state,
```
Comment (psc-g) [inline lifting_veil/agents/dqn_agent_new.py:319]: you still need to pass `self.optimizer` as the 4th parameter

File: lifting_veil/agents/rainbow_agent_new.py [hunk 4]
```diff
@@ -24,13 +24,18 @@
 import jax.numpy as jnp
 import numpy as onp
 import tensorflow as tf
+import optax
 
+#@functools.partial(jax.jit, static_argnums=(0, 10, 11))
```
Comment (psc-g) [inline lifting_veil/agents/rainbow_agent_new.py]: remove commented out lines?

File: lifting_veil/agents/rainbow_agent_new.py [hunk 5]
```diff
@@ -24,13 +24,18 @@
...
+#          terminals, loss_weights, support, cumulative_gamma, double_dqn, rng):
 
-@functools.partial(jax.jit, static_argnums=(0, 10, 11))
-def train(network_def, target_params, optimizer, states, actions, next_states, rewards,
+@functools.partial(jax.jit, static_argnums=(0, 3, 12, 13))
+def train(network_def, online_params, target_params, optimizer, optimizer_state, states, actions, next_states, rewards,
           terminals, loss_weights, support, cumulative_gamma, double_dqn, rng):
   """Run a training step."""
-  online_params = optimizer.target
+  """Run a training step."""
```
Comment (psc-g) [inline lifting_veil/agents/rainbow_agent_new.py]: remove?

File: lifting_veil/agents/rainbow_agent_new.py [hunk 6]
```diff
@@ -61,8 +66,10 @@ def q_target_online(state):
   
   # Get the unweighted loss without taking its mean for updating priorities.
   (mean_loss, loss), grad = grad_fn(online_params, rng3, target, loss_weights)
+  updates, optimizer_state = optimizer.update(grad, optimizer_state)
+  online_params = optax.apply_updates(online_params, updates)
   optimizer = optimizer.apply_gradient(grad)
```
Comment (psc-g) [inline lifting_veil/agents/rainbow_agent_new.py]: remove

File: lifting_veil/agents/rainbow_agent_new.py [hunk 7]
```diff
@@ -229,10 +236,12 @@ def __init__(self,
...
-    online_network_params = self.network_def.init(rng, x=self.state, support=self._support, rng=self._rng)
-    optimizer_def = dqn_agent.create_optimizer(self._optimizer_name)
-    self.optimizer = optimizer_def.create(online_network_params)
-    self.target_network_params = copy.deepcopy(online_network_params)
+    self.online_params = self.network_def.init(rng, x=self.state,
+                                               support=self._support, 
+                                               rng=self._rng)
+    self.optimizer = dqn_agent.create_optimizer(self._optimizer_name)
+    self.optimizer_state = self.optimizer.init(self.online_params)
+    self.target_network_params = copy.deepcopy(self.online_params)
```
Comment (psc-g) [inline lifting_veil/agents/rainbow_agent_new.py:238]: you don't need `copy.deepcopy`

File: lifting_veil/agents/rainbow_agent_new.py [hunk 8]
```diff
@@ -322,10 +331,12 @@ def _train_step(self):
         else:
           loss_weights = jnp.ones(self.replay_elements['state'].shape[0])
 
-        self.optimizer, loss, mean_loss = train(
+        (self.optimizer_state, self.online_params, loss, mean_loss) = train(
```
Comment (psc-g) [inline lifting_veil/agents/rainbow_agent_new.py]: you don't need the parentheses on the left-hand side. i may have used them in the dopamine code base because there was a newline in the middle.

**Review by psc-g — CHANGES_REQUESTED** [review]
almost there johan, a few comments below that need to be fixed before it runs.

would you mind testing it out before sending it out? i don't think it will run as it is right now.

File: lifting_veil/agents/rainbow_agent_new.py (hunk 8 above)
Comment (johanobandoc) [reply lifting_veil/agents/rainbow_agent_new.py]: Perfect! Thanks.

File: lifting_veil/agents/rainbow_agent_new.py (hunk 7 above)
Comment (johanobandoc) [reply lifting_veil/agents/rainbow_agent_new.py:238]: You are using it here: https://github.com/google/dopamine/blob/dabc84b1013b7ced981f189b2bb161e7fa94d31a/dopamine/jax/agents/rainbow/rainbow_agent.py#L283

Should I remove it?

File: lifting_veil/agents/rainbow_agent_new.py (hunk 6 above)
Comment (johanobandoc) [reply lifting_veil/agents/rainbow_agent_new.py]: I think I should remove just line 71.
You are using line 60 and 70: https://github.com/google/dopamine/blob/dabc84b1013b7ced981f189b2bb161e7fa94d31a/dopamine/jax/agents/rainbow/rainbow_agent.py#L87

File: lifting_veil/agents/dqn_agent_new.py (hunk 1 above)
Comment (johanobandoc) [reply lifting_veil/agents/dqn_agent_new.py]: Yes, perfect!

File: lifting_veil/agents/dqn_agent_new.py (hunk 2 above)
Comment (johanobandoc) [reply lifting_veil/agents/dqn_agent_new.py:276]: You are still using it:
https://github.com/google/dopamine/blob/dabc84b1013b7ced981f189b2bb161e7fa94d31a/dopamine/jax/agents/dqn/dqn_agent.py#L328

Should I remove it?

File: lifting_veil/agents/dqn_agent_new.py (hunk 3 above)
Comment (johanobandoc) [reply lifting_veil/agents/dqn_agent_new.py:319]: Yes, perfect!

File: lifting_veil/agents/rainbow_agent_new.py (hunk 4 above)
Comment (johanobandoc) [reply lifting_veil/agents/rainbow_agent_new.py]: Yes, perfect!

File: lifting_veil/agents/rainbow_agent_new.py (hunk 5 above)
Comment (johanobandoc) [reply lifting_veil/agents/rainbow_agent_new.py]: Yes, perfect!

File: lifting_veil/agents/dqn_agent_new.py (hunk 2 above)
Comment (psc-g) [reply lifting_veil/agents/dqn_agent_new.py:276]: yeah, you can remove it. i should also remove it from the dopamine code.

File: lifting_veil/agents/rainbow_agent_new.py (hunk 6 above)
Comment (psc-g) [reply lifting_veil/agents/rainbow_agent_new.py]: yes, the comment was for line 71. you need to keep 69 and 70.

File: lifting_veil/agents/rainbow_agent_new.py (hunk 7 above)
Comment (psc-g) [reply lifting_veil/agents/rainbow_agent_new.py:238]: same comment as above.

**Comment by psc-g** [comment]
thanks johan. i don't see the recent changes you made, though.

File: lifting_veil/agents/dqn_agent_new.py (hunk 2 above)
Comment (psc-g) [reply lifting_veil/agents/dqn_agent_new.py:276]: reminder that you can remove `copy.deepcopy`

File: lifting_veil/agents/rainbow_agent_new.py (hunk 7 above)
Comment (psc-g) [reply lifting_veil/agents/rainbow_agent_new.py:238]: reminder that you can remove `copy.deepcopy`

**Review by psc-g — APPROVED** [review]

**Review by joaogui1 — APPROVED** [review]
LGTM!
